dgl_sparse/src/matmul.cc

```diff
@@ -98,6 +98,53 @@ torch::Tensor SDDMMNoAutoGrad(
   return ret;
 }
 
+torch::Tensor BroadcastOpNoAutoGrad(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor e,
+    torch::Tensor v, const std::string& op) {
+  const int64_t out_row = sparse_mat->nnz();
+  const std::vector<int64_t> shape({out_row, e.size(1)});
+  auto ret = torch::zeros(shape, e.options());
+
+  auto dgl_e = TorchTensorToDGLArray(e);
+  auto dgl_v = TorchTensorToDGLArray(v);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+
+  // The format for calculation will be chosen in the following order: CSR,
+  // COO. CSR is created if the sparse matrix only has CSC format.
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
+    // sparse_mat->CSRPtr() will implicitly convert CSC to CSR format if CSR
+    // does not exist.
+    auto csr = CSRToOldDGLCSR(sparse_mat->CSRPtr());
+    aten::CSRSDDMM(
+        op.c_str(), csr, dgl_e, dgl_v, dgl_ret, 1 /* Lhs target: e */,
+        0 /* rhs target: u due to transpose */);
+  } else {  // COO
+    auto coo = COOToOldDGLCOO(sparse_mat->COOPtr());
+    aten::COOSDDMM(
+        op.c_str(), coo, dgl_e, dgl_v, dgl_ret, 1 /* Lhs target: e */,
+        0 /* rhs target: u due to transpose */);
+  }
+  return ret;
+}
+
+torch::Tensor BroadcastSubNoAutoGrad(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor e,
+    torch::Tensor v) {
```

Comment:
I think a better signature just needs a sparse matrix and a dense tensor. The non-zero values are naturally carried in the sparse matrix. The return will also be a sparse matrix.

Reply:
updated